src/Magento/FunctionalTestingFramework/Test/Util/ActionMergeUtil.php

```diff
@@ -217,8 +217,32 @@ private function mergeActions($parsedSteps)
      */
     private function insertWaits()
     {
+        # This list is sorted alphabetically. Please maintain that order when modifying.
+        $blacklist = [
+            "assertArrayHasKey", "assertArrayIsSorted", "assertArrayNotHasKey", "assertArraySubset", "assertContains",
+            "assertCount", "assertElementContainsAttribute", "assertEmpty", "assertEquals", "assertFalse",
+            "assertFileExists", "assertFileNotExists", "assertGreaterOrEquals", "assertGreaterThan",
+            "assertGreaterThanOrEqual", "assertInstanceOf", "assertInternalType", "assertIsEmpty", "assertLessOrEquals",
+            "assertLessThan", "assertLessThanOrEqual", "assertNotContains", "assertNotEmpty", "assertNotEquals",
+            "assertNotInstanceOf", "assertNotNull", "assertNotRegExp", "assertNotSame", "assertNull", "assertRegExp",
+            "assertSame", "assertStringStartsNotWith", "assertStringStartsWith", "assertTrue", "createData",
+            "deleteData", "dontSee", "dontSeeCheckboxIsChecked", "dontSeeCookie", "dontSeeCurrentUrlEquals",
+            "dontSeeCurrentUrlMatches", "dontSeeElement", "dontSeeElementInDOM", "dontSeeInCurrentUrl",
+            "dontSeeInField", "dontSeeInFormFields", "dontSeeInPageSource", "dontSeeInSource", "dontSeeInTitle",
+            "dontSeeJsError", "dontSeeLink", "dontSeeOptionIsSelected", "expectException", "formatMoney",
+            "generateDate", "getData", "grabAttributeFrom", "grabCookie", "grabFromCurrentUrl", "grabMultiple",
+            "grabPageSource", "grabTextFrom", "grabValueFrom", "magentoCLI", "parseFloat", "see",
+            "seeCheckboxIsChecked", "seeCookie", "seeCurrentUrlEquals", "seeCurrentUrlMatches", "seeElement",
+            "seeElementInDOM", "seeInCurrentUrl", "seeInField", "seeInFormFields", "seeInPageSource", "seeInPopup",
+            "seeInSource", "seeInTitle", "seeLink", "seeNumberOfElements", "seeOptionIsSelected", "switchToIFrame",
+            "switchToNextTab", "switchToPreviousTab", "switchToWindow", "updateData", "wait", "waitForAjaxLoad",
+            "waitForElement", "waitForElementChange", "waitForElementNotVisible", "waitForElementVisible", "waitForJS",
+            "waitForLoadingMaskToDisappear", "waitForPageLoad", "waitForPwaElementNotVisible",
+            "waitForPwaElementVisible", "waitForText"
+        ];
+
         foreach ($this->orderedSteps as $step) {
-            if ($step->getTimeout()) {
+            if ($step->getTimeout() && !in_array($step->getType(), $blacklist)) {
```

Comment:
You may want to consider changing the above array to be `"actionType"=>0"` so they are keyed instead of valued (or `array_flip()` the human readable array above) and check on them via `isset($blacklist[$step->getType()])`

I ran a profiler a couple times and doing index checking is consistently close to twice as fast. (avg 3,600ms vs 1,700ms)

Reply:
Done